app/controllers/group_controller.rb

```diff
@@ -0,0 +1,50 @@
+class GroupController < ApplicationController
+  protect_from_forgery with: :null_session
+
+  def index
+  end
+
+  def show
+    @group = Device.where(group: params[:group_id])
+  end
+
+  def raw_data
+    kg = params[:unit] == 'kg_per_year'
+    devices = Device.select('devices.*, max / (cpus * cores_per_cpu) AS max_per_core')
+    .group(:group)
+    .order(:max_per_core)
+    response = {}.tap do |res|
+      max_device = devices.last
+      res[:max_main] = max_device&.max_per_core * (kg ? 8.76 : 1)
+      res[:header] = {}.tap do |h|
+        h[:user] = 'Group'
+        h[:platform] = 'Platform'
+        h[:location] = 'Location'
+        h[:core_number] = 'No. cores'
+        h[:ram] = 'RAM (GB)'
+        h[:main] = "Carbon emissions per core at full load (CO2eq/#{kg ? 'year' : 'hour'})"
+      end
+      rank = 1
+      res[:groups] = devices
+      .group_by(&:max_per_core)
+      .values.flat_map do |dev_group|
+        current_rank = rank
+        rank += dev_group.size
+        dev_group.map do |dev|
+          {}.tap do |new_dev|
+            new_dev[:rank] = current_rank
+            new_dev[:user] = dev.pretty_group
+            new_dev[:platform] = dev.platform
+            new_dev[:location] = dev.location
+            new_dev[:core_number] = dev.cpus * dev.cores_per_cpu
+            new_dev[:ram] = dev.ram_units * dev.ram_capacity_per_unit
+            new_dev[:main] = "#{signif(dev.max_per_core * (kg ? 8.76 : 1), 3)}#{kg ? 'k' : ''}g"
```

Comment:
I found the bug with the bar lengths - this `main` value is being used to calculate the bar length in the js so making it into a string causes the calculation to break. Separating the value and units into two fields should fix it

Reply:
Also, what is the `(kg ? 8.76 : 1)` for?

Reply:
That's for converting between g and kg. `kg` is a boolean which is `true` if the leaderboard is in `kg/year` mode. Converting from g/hour to kg/year, the formula is `gramsperhour * 24 * 365 / 1000` which is equivalent to `gramsperhour * 8.76`.